ssl/ssl_test.cc

```diff
@@ -1014,6 +1089,318 @@ static const HybridHandshakeTest kHybridHandshakeTests[] = {
+    TLS1_3_VERSION,
+    "SecP256r1MLKEM768:prime256v1",
+    TLS1_3_VERSION,
+    SSL_GROUP_SECP256R1_MLKEM768,
+    false,
+  },
+
+  // The client will send key shares [x25519, SecP256r1MLKEM768]. The
+  // server will negotiate x25519 since the client listed it as its first
+  // preference, even though it supports SecP256r1MLKEM768.
+  {
+    "x25519:prime256v1:SecP256r1MLKEM768",
+    TLS1_3_VERSION,
+    "prime256v1:x25519:SecP256r1MLKEM768",
+    TLS1_3_VERSION,
+    SSL_GROUP_X25519,
+    false,
+  },
+
+  // The client will send key shares [SecP256r1MLKEM768, x25519].
+  // The server will negotiate SecP256r1MLKEM768 since the client listed
+  // it as its first preference.
+  {
+    "SecP256r1MLKEM768:x25519:prime256v1",
+    TLS1_3_VERSION,
+    "prime256v1:x25519:SecP256r1MLKEM768",
+    TLS1_3_VERSION,
+    SSL_GROUP_SECP256R1_MLKEM768,
+    false,
+  },
+
+  // In the supported_groups extension, the client will indicate its
+  // preferences, in order, as [SecP256r1MLKEM768, X25519MLKEM768,
+  // x25519, prime256v1]. From those groups, it will send key shares
+  // [SecP256r1MLKEM768, x25519]. The server supports, and receives a
+  // key share for, x25519. However, when selecting a mutually supported group
+  // to negotiate, the server recognizes that the client prefers
+  // X25519MLKEM768 over x25519. Since the server also supports
+  // X25519MLKEM768, but did not receive a key share for it, it will
+  // select it and send an HRR. This ensures that the client's highest
+  // preference group will be negotiated, even at the expense of an additional
+  // round-trip.
+  //
+  // In our SSL implementation, this situation is unique to the case where the
+  // client supports both ECC and hybrid/PQ. When sending key shares, the
+  // client will send at most two key shares in one of the following ways:
+
+  // (a) one ECC key share - if the client supports only ECC;
+  // (b) one PQ key share - if the client supports only PQ;
+  // (c) one ECC and one PQ key share - if the client supports ECC and PQ.
+  //
+  // One of the above cases will be true irrespective of how many groups
+  // the client supports. If, say, the client supports four ECC groups
+  // and zero PQ groups, it will still only send a single ECC share. In cases
+  // (a) and (b), either the server supports that group and chooses to
+  // negotiate it, or it doesn't support it and sends an HRR. Case (c) is the
+  // only case where the server might receive a key share for a mutually
+  // supported group, but chooses to respect the client's preference order
+  // defined in the supported_groups extension at the expense of an additional
+  // round-trip.
+  {
+    "SecP256r1MLKEM768:X25519MLKEM768:x25519:prime256v1",
+    TLS1_3_VERSION,
+    "X25519MLKEM768:prime256v1:x25519",
+    TLS1_3_VERSION,
+    SSL_GROUP_X25519_MLKEM768,
+    true,
+  },
+
+  // Like the previous case, but the client's prioritization of ECC and PQ
+  // is inverted.
```

Comment:
Interesting, so if the client supports PQ, but doesn't put it first the server won't do a HRR to get the PQ key share? Does this follow s2n's behavior?

Reply:
This is different than s2n's behavior, but matches the existing long term Openssl/Boringssl/AWS-LC/LibreSSL behavior to use the client's preference order no matter what by default.

s2n's negotiation logic is more complex "PQ > 1-RTT > ServerPreference" is used, where PQ algorithms are preferred over non-PQ, then algorithms negotiable in 1-RTT are preferred over 2-RTT, and finally the server-side preference order is used.